<a href="https://colab.research.google.com/github/emmannuelltr/Analitica-de-Negocios/blob/main/carpeta1/Untitled0.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##**1. Descripción del Problema**

Una institución de salud busca implementar un modelo de árbol de decisión para mejorar el diagnóstico preventivo y el pronóstico de enfermedades coronarias en sus pacientes. Para este proceso vamos a utilizar las variables extraídas de la base de datos clínica:

Edad: Indica el número de años que posee el paciente, siendo un factor de riesgo biológico fundamental.

Systolic (Presión Sistólica): Representa la presión máxima que ejerce el corazón cuando late (medida en mmHg).

Diastolic (Presión Diastólica): Indica la presión en las arterias cuando el corazón reposa entre latidos (medida en mmHg).

BMI (Índice de Masa Corporal): Es una medida de la grasa corporal basada en la relación entre el peso y la estatura del individuo.

In [ ]:
#Se procede con la carga de las librerías de trabajo}
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import confusion_matrix

In [ ]:
#Se procede con la carga de datos de trabajo
nxl="/content/2. HeartDisease.xlsx"
XDB=pd.read_excel(nxl, sheet_name=0)

XD=XDB.iloc[:, [1, 2, 3, 4]]
yd=XDB.iloc[:, [0]]
display(XDB)

,Disease,Age,Systolic,Diastolic,BMI
0,0,44,112,111,17
1,1,55,128,90,27
2,1,47,131,94,26
3,0,31,151,104,17
4,1,65,148,117,17
...,...,...,...,...,...
175,0,43,119,76,25
176,1,63,107,113,40
177,0,47,116,61,27
178,0,70,110,88,24


In [ ]:
#Se procede con la implementaciòn del modelo de árbol
from sklearn.metrics import confusion_matrix

mar = DecisionTreeClassifier(criterion="entropy", max_depth=3) # Ajustado a tu reto de 3 niveles
mar.fit(XD, yd)

ydp = mar.predict(XD)   # Pronóstico del modelo sobre los datos históricos

# 2. Se construye la matriz de confusión
cm = confusion_matrix(yd, ydp)
VN = cm[0,0]; FP = cm[0,1]; FN = cm[1,0]; VP = cm[1,1]
print("Matriz de confusión: \n", cm)

# 3. Métricas de desempeño (Basadas en tu ejemplo)
Ex = (VP + VN) / len(XD) # Exactitud: Comportamiento general del modelo
print("Exactitud: ", Ex)

Sen = VP / (VP + FN) # Sensibilidad: Capacidad para detectar enfermos (Disease=1)
print("Sensibilidad (Recall): ", Sen)

Sp = VN / (VN + FP) # Especificidad: Capacidad para detectar personas sanas (Disease=0)
print("Especificidad: ", Sp)

Prec = VP / (VP + FP) # Precisión: De los pronosticados como enfermos, ¿cuántos lo están?
print("Precisión: ", Prec)

PreNeg = VN / (VN + FN) # Precisión Negativa: De los pronosticados como sanos, ¿cuántos lo están?
print("Precisión Negativa: ", PreNeg)

Matriz de confusión: 
 [[64  9]
 [33 74]]
Exactitud:  0.7666666666666667
Sensibilidad (Recall):  0.6915887850467289
Especificidad:  0.8767123287671232
Precisión:  0.891566265060241
Precisión Negativa:  0.6597938144329897


In [ ]:
#DESPLIEGUE ARBOL DE DECISIÓN

In [ ]:
from sklearn.tree import export_graphviz #exporta datos a un gráfico
from pydotplus import graph_from_dot_data #crea un gráfico a partir de los datos

vc=["Age","Systolic","Diastolic","BMI"] #Título del árbol
dot_data=export_graphviz(mar, feature_names=vc) #Exportar de números a gráfico
graph=graph_from_dot_data(dot_data) #Crea el gráfico
graph.write_png("Arbol.png")


True

##**ANALÍSIS DE RESULTADO**
De la base de datos se puede observar un total de 73 personas que poseen la categoría de sanos, mientras que para la categoría de enfermos la matriz cuenta con un total de 107 personas con enfermedad cardiaca. De los 180 datos totales de la base de entrenamiento, el modelo pronosticó correctamente un total de 138. El modelo logra identificar en un 87,67% a las personas sanas (64/73), mientras que el modelo logró identificar en un 69,15% a las personas enfermas (74/107).

Con respecto a las métricas el modelo logró una exactitud de 76,66% lo que indica el buen comportamiento general del modelo frente a la clasificación de salud en las dos categorías de diagnóstico. Se destacan la especificidad y la precisión las cuales lograron valores por encima del 87% y 89% respectivamente, lo que supera el límite inferior definido para este tipo de modelos de clasificación el cual se ubica en el 75%.

De acuerdo con el árbol de decisión se destaca un nodo puro el cual posee la siguiente decisión, si una persona cumple con esta SI Age<=54 and Systolic<=129 and BMI<=25, tendrá una probabilidad de estar sano del 100%. Es importante destacar que los nodos puros poseen una entropía=0 (El modelo diferencia muy bien a las personas sin riesgo).

A pesar de la no existencia de más nodos puros, se destacan los nodos extremos o las reglas de salud extremas. Se destaca una segunda regla que logra un porcentaje de diagnóstico sano elevado, y la cual posee la siguiente estructura: Si Age<=54 and Systolic<=129 and BMI>25 and Diastolic<=80. Por su parte, el nodo derecho posee un porcentaje de enfermedad del 89%, en donde la regla se define: si Age>54 and Systolic>130 and BMI>30 y la presión sistólica es mayor a 140 (9/74).

In [ ]:
from sklearn.tree import export_graphviz #exporta datos a un gráfico
from pydotplus import graph_from_dot_data #crea un gráfico a partir de los datos

# Títulos de las variables de tu archivo Heart Disease
vc=["Age","Systolic","Diastolic","BMI"]

# Exportar de números a gráfico
dot_data=export_graphviz(mar, feature_names=vc, class_names=['Sano', 'Enfermo'], filled=True, rounded=True)

# Crea el gráfico
graph=graph_from_dot_data(dot_data)

# Guarda el gráfico en formato PDF
graph.write_pdf("Arbol.pdf")

True